airflow/jobs.py

```diff
@@ -1891,118 +1925,92 @@ def _manage_executor_state(self, started):
                     self.logger.error(msg)
                     ti.handle_failure(msg)
 
-    def _execute(self):
+    @provide_session
+    def _tasks_instances_for_run_date(self, run_date, session=None):
```

Comment:
I think returning two things is a little bit unclean here. I think it would be fine to just duplicate the query to get the dagrun in the caller function if it's necessary. Alternatively we could make dagrun an optional parameter for this function (so we would only compute dagrun if the caller does not specify it), but I think performance isn't that important for backfill jobs since they are adhoc (except for subdag operators which are going to go away eventually).

Also "tasks_instances" should be "task_instances"